hail/python/hailtop/batch/batch_pool_executor.py

```diff
@@ -344,6 +345,7 @@ async def async_submit(self,
         dill.dump(functools.partial(unapplied, *args, **kwargs), pipe, recurse=True)
         pipe.seek(0)
         pickledfun_gcs = self.inputs + f'{name}/pickledfun'
+        await self.gcs_fs.makedirs(os.path.dirname(pickledfun_gcs), exist_ok=True)
```

Comment:
I think I'm missing something about this code. It shouldn't be necessary for an object store and certainly isn't necessary for GCS. We know `gcs_fs` is a GCS client and any future evolution of this code would at most make it a generic object store client.

I think what I'm generally reacting to is: I don't expect to have to create directories in an object store; therefore, when I read this line, I wonder what hidden subtle issues are addressed by this line. I can't seem to think of any?

Reply:
I suppose the alternative is that we, as Hail team, make it a rule that we will *always* use `makedirs` on any file stored in GCS. I don't quite see the benefit of that, but consistently doing it everywhere at least avoids my confusion (because it will be common and therefore not surprising).

Reply:
The issue is the file system is actually a RouterFileSystem. I'll clean up the naming of that variable. This is needed for the python job to work on the local backend. The local backend needs makedirs to be called to write the file to the correct place. We can check explicitly whether `pickledfun_gcs` is local or remote, but that seems to defeat the purpose of the router file system at least to me.

Reply:
bump.

Reply:
This variable always a GoogleStorageAsyncFS (line 137). The BatchPoolExecutor doesn't really make sense locally. Locally, it's just a slower ProcesPoolExecutor. If we wanted to do this, I'd create the directories once in the constructor.

The BatchPoolExecutor already asserts that its using a ServiceBackend, so there's no concern that someone might try to uses the BatchPoolExecutor on the local backend.